### ISSUE TITLE
Increase the number of scrypt rounds

### ISSUE THREAD

**Issue by jvoisin** [syncthing/syncthing#8193]
The current number of scrypt iterations used by syncthing is [32768](https://github.com/syncthing/syncthing/blob/main/lib/protocol/encryption.go#L494),
but this is the [recommended value](
https://pkg.go.dev/golang.org/x/crypto/scrypt) for interactive logins.
As the security maintainer for go's crypto module said in [his blogpost
from 2017]( https://words.filippo.io/the-scrypt-parameters/ ):

```
    interactive logins: 2^15 — 1 << 15 — 32 768 — 86ms
    file encryption: 2^20 — 1 << 20 — 1 048 576 — 3802ms
```

and this was from 5 years ago, so 32k is dangerously small, so you might want to increase it, and maybe store it in the encrypted files' metadata, so that it can be modified in the future.

It would also make sense to stretch the passphrase a bit *before* deriving it with `FolderID`: better to do an expensive computation once, then a couple of small ones than to do a lot of heavy ones.

**Comment by greatroar** [comment]
Yes, it would be better if this were configurable (with 15 the minimum). Note that scrypt uses (1<<N)*2KiB memory, though, which is 2GiB at N=20.

**Comment by calmh** [comment]
There will be some technical details to sort out in terms of making sure that the same setting is used among devices sharing a given folder etc.

**Comment by imsodin** [comment]
There's already a token passed around that needs to match - this might very well cause a failure already (with a then misleading message about mismatching passwords).
Also 2G? o.O All those old phone and other micro-device users will be furious.

**Comment by jvoisin** [comment]
The derivation only needs to be done on the sender's end. Large memory requirement is an [explicit goal of scrypt]( https://en.wikipedia.org/wiki/Scrypt), and `1<<15` (aka 86ms 5 years ago) is *really* low :/

An other solution would be to move to [Argon2id]( https://en.wikipedia.org/wiki/Argon2 ).

**Comment by bt90** [comment]
I share @imsodin opinion. 2GB memory usage is an absolute dealbreaker IMHO.

**Comment by tomasz1986** [comment]
> Also 2G? o.O All those old phone and other micro-device users will be furious.

Not only that. I've worked in educational institutions and now help a friend with administering Syncthing at a school. Desktop computers with 2GB of RAM or even less are still very common there. Schools generally use computers until they break, and also due to constant financial constraints they tend to always go for low-end hardware even when buying new ones.

**Comment by rdebath** [comment]
I was wandering past and noticed your conversation ...

> Also 2G? o.O All those old phone and other micro-device users will be furious.
> ... Desktop computers with 2GB of RAM or even less ...

Please, do remember scrypt is a "defence in depth" or mitigation strategy. For it to have any effect the user has to have, at least, already chosen a poor password. In particular there is no way that the count could be considered "dangerously small" because it's reasonable to estimate it's "best world" effect is zero.

But, because the "real world" isn't perfect, this is still a good thing to have, though the allowable cost (ie the number of rounds) is driven almost entirely by how much of a problem it is. Mostly, how long does a user on an expected client have to wait for the process to complete (not completing is bad!). 

So for the attached blog; the author uses their laptop as the expected device and the normal sub-second and five second delays for the patience of the user.

Syncthing apparently has different requirements for these limits (eg "phones", "OldCrap" and "WeirdCrap") .

PS: Argon2 series ... a quick skim reveals that this is really, really, new. The Alwen/Blocki attack is worrying, though as it only applies to the side-channel resistant variant this is probably not a severe issue as yet. Also you'd really need to check performance on ARM (etc) as it's supposed to be really slow for 32bit CPUs and other non-AMD64 machines.

**Comment by greatroar** [comment]
I was off by a factor of two: it's 1GiB, not 2GiB, for N=20.

**Comment by calmh** [comment]
(Relevant: https://forum.syncthing.net/t/problem-with-working-with-a-cluster-of-multiple-devices/19073)

**Comment by bt90** [comment]
Argon2 - namely the argon2d and argon2id variants - is the new default for multiple password managers. I think we should follow here and replace scrypt.
